File: LeetSync/977-distinct-subsequences-ii/distinct-subsequences-ii.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        MOD = 10**9 + 7
        ends_with = [0] * 26
        
        for ch in s:
            idx = ord(ch) - ord('a')
            ends_with[idx] = (1 + sum(ends_with)) % MOD
            
        return sum(ends_with) % MOD

File: LeetSync/977-distinct-subsequences-ii/distinct-subsequences-ii.cpp
class Solution {
public:
    int distinctSubseqII(string s) {
        const int MOD = 1e9 + 7;
        long long ends_with[26] = {0};
        
        for (char c : s) {
            int idx = c - 'a';
            long long new_subsequences_ending_in_c = 1;
            
            for (int i = 0; i < 26; ++i) {
                new_subsequences_ending_in_c = (new_subsequences_ending_in_c + ends_with[i]) % MOD;
            }
            
            ends_with[idx] = new_subsequences_ending_in_c;
        }
        
        long long total = 0;
        for (int i = 0; i < 26; ++i) {
            total = (total + ends_with[i]) % MOD;
        }
        
        return total;
    }
};